### PR TITLE
Issue/13370 retry button failed uploads

### LINKED ISSUE #13370: Add retry button to failed uploads within the fail uploads screen

### PR DESCRIPTION
**Pull request by JorgeMucientes**
Closes: #13370

### Description

The main change of this PR is to add the retry functionality to `MediaUploadErrorListFragment`. However, I found the code of this screen fairly outdated and since it is very likely we are going to apply more changes to this screen in itearation II of the image upload improvements project I also applied the following changes: 
- Migrated the UI to compose, simplifying the code and getting rid of the adapters and xml code. This will make it easier in future iterations to add changes. 
- Added retry feature
- Added unit tests to screen logic

There are a few known issues that should be addressed in a different PR: 
- The upload error screen is accessible only once. Once opened when navigating back, the snackbar providing the CTA is gone. I think this screen should be available all the time. 
- If we kill the app while the images are being uploaded, when tapping on the error notification to open the uploads error screen, the retry fuction won't work: 

https://github.com/user-attachments/assets/ef449f3a-061c-47e8-a17a-74153f0f9a11

The reason for this, is that once the app is killed we loose permission to access any uri from the photPicker leading to the following error: 

```
Can't download the file at: content://media/picker/0/com.android.providers.media.photopicker/media/1000000096
java.lang.SecurityException: Calling uid ( 10200 ) does not have permission to access picker uri: content://media/picker/0/com.android.providers.media.photopicker/media/1000000096
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
```
We might need to use [takePersistableUriPermission](https://developer.android.com/reference/android/content/ContentResolver#takePersistableUriPermission(android.net.Uri,%20int)) to solve the issue. In any case, it is a corner case so I think the PR can be merged without this for now. 

### Testing information

To test the following I suggest using API faker to make the media upload requests fail. You can apply the following configuration to More menu > Settings > Developer Options > API faker: 

 

Once API faker is enabled: 
1. Open product detail
2. Open the product images screen
3. Tap `Add photos`
4. Tap `Choose from device`
5. Select a few photos and let the uploads fail.
6. Open the error screen and tap retry. See how the upload is retried again until it fails again and is showed again on the error list. 

Repeat the steps above and test the following: 

1. Try to upload a single image 
2. Open the error screen
3. Tap retry button and see how the screen is closed automatically and app navigates back to product images screen where the retry upload image should show loading. 

https://github.com/user-attachments/assets/54c91cef-af16-4be7-b5a8-a6c55b53bb8a

### The tests that have been performed

The above. 

### Images/gif

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListViewModel.kt [hunk 1]
```diff
@@ -2,68 +2,82 @@ package com.woocommerce.android.ui.media
...
 import dagger.hilt.android.lifecycle.HiltViewModel
 import kotlinx.coroutines.flow.filter
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
 import kotlinx.parcelize.Parcelize
 import javax.inject.Inject
 
 @HiltViewModel
 class MediaUploadErrorListViewModel @Inject constructor(
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListViewModel.kt:23]: Refactored most of the logic from this class, so it can be checked as a new file.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                navigationIcon = Icons.AutoMirrored.Filled.ArrowBack,
+            )
+        },
+        content = {
+            MediaUploadErrorList(
+                errors = state.uploadErrorList,
+                onRetryClicked = onRetryClicked,
+                modifier = Modifier
+                    .background(MaterialTheme.colors.surface)
+                    .padding(it)
```
Comment (irfano) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt]: ❓Question: This `Scaffold` does not have any inner padding, so `it` is 0 here and `.padding(it)` has no effect. Did you add this intentionally as a best practice or for a specific reason?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    ) {
+        AsyncImage(
+            model = ImageRequest.Builder(LocalContext.current)
+                .data(error.localUri)
+                .crossfade(true)
+                .build(),
+            fallback = painterResource(R.drawable.ic_product),
+            placeholder = painterResource(R.drawable.ic_product),
+            error = painterResource(R.drawable.ic_product),
+            contentDescription = "",
```
Comment (irfano) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt]: Let's use `R.string.product_image_content_description` as the `contentDescription`.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+@ExperimentalCoroutinesApi
+class MediaUploadErrorListViewModelTest : BaseUnitTest() {
+    companion object {
+        private const val REMOTE_PRODUCT_ID = 1L
+        private val SOME_UPLOAD_ERRORS = arrayOf(
+            ProductImageUploadData(
+                localUri = "file://test1.jpg",
+                remoteProductId = REMOTE_PRODUCT_ID,
+                uploadStatus = MediaFileUploadHandler.UploadStatus.Failed(
```
Comment (irfano) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListViewModelTest.kt]: Please consider fixing these IDE warnings about redundant qualifier names.
image

**Review by irfano — APPROVED** [review]
Great job refactoring this screen, @JorgeMucientes! I've added a small request related content description and 2 comments, but I'm approving. It works as expected! 👍🏻

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt (hunk 2 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt]: Good question @irfano . It is recommended to pass the padding from the scafold to the childview as there could be cases where it might not be 0. Removing that padding will lead to a Lint error "padding not used". 
Screenshot 2025-01-28 at 17 08 16

Comment (irfano) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/media/MediaUploadErrorListScreen.kt]: TIL, Thank you!

**Review by irfano — APPROVED** [review]
LGTM! 👍🏻
